// slate-edit-code/dist/handlers/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.onKeyDown = exports.onPaste = exports.onSelectAll = exports.onBackspace = exports.onModEnter = exports.onEnter = exports.onShiftTab = exports.onTab = undefined;

var _onTab = require('./onTab');

var _onTab2 = _interopRequireDefault(_onTab);

var _onShiftTab = require('./onShiftTab');

var _onShiftTab2 = _interopRequireDefault(_onShiftTab);

var _onEnter = require('./onEnter');

var _onEnter2 = _interopRequireDefault(_onEnter);

var _onModEnter = require('./onModEnter');

var _onModEnter2 = _interopRequireDefault(_onModEnter);

var _onBackspace = require('./onBackspace');

var _onBackspace2 = _interopRequireDefault(_onBackspace);

var _onSelectAll = require('./onSelectAll');

var _onSelectAll2 = _interopRequireDefault(_onSelectAll);

var _onPaste = require('./onPaste');

var _onPaste2 = _interopRequireDefault(_onPaste);

var _onKeyDown = require('./onKeyDown');

var _onKeyDown2 = _interopRequireDefault(_onKeyDown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.onTab = _onTab2.default;
exports.onShiftTab = _onShiftTab2.default;
exports.onEnter = _onEnter2.default;
exports.onModEnter = _onModEnter2.default;
exports.onBackspace = _onBackspace2.default;
exports.onSelectAll = _onSelectAll2.default;
exports.onPaste = _onPaste2.default;
exports.onKeyDown = _onKeyDown2.default;
